### ISSUE TITLE
pylint - false positive E0401:import-error

### ISSUE THREAD

**Issue by jacleggmsr** [microsoft/vscode-pylint#219]
Type: <b>Bug</b>

Something odd is going on in the new pylint extension. I'm not expert enough to pinpoint it.  The extension gives import errors for all modules from my current repo despite them being on my PYTHONPATH.    Running pylint via command line and omitting the "--from-stdin" flag gives no import errors and the code runs fine so I'm pretty sure this is something to do with the extension.

Extension version: 2022.6.0
VS Code version: Code 1.73.1 (6261075646f055b99068d3688932416f2346dd3b, 2022-11-09T04:27:29.066Z)
OS version: Windows_NT x64 10.0.22621
Modes:
Sandboxed: No

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-1065G7 CPU @ 1.30GHz (8 x 1498)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: disabled_off<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_renderer: enabled_on<br>video_decode: enabled<br>video_encode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled<br>webgpu: disabled_off|
|Load (avg)|undefined|
|Memory (System)|15.60GB (6.28GB free)|
|Process Argv|--crash-reporter-id d25bcfbe-00a5-4538-a621-b814c5cf1d5d|
|Screen Reader|no|
|VM|0%|
</details><details>
<summary>A/B Experiments</summary>

```
vsliv368cf:30146710
vsreu685:30147344
python383:30185418
vspor879:30202332
vspor708:30202333
vspor363:30204092
vswsl492cf:30256860
vslsvsres303:30308271
pythonvspyl392:30443607
vserr242:30382549
pythontb:30283811
vsjup518:30340749
pythonptprofiler:30281270
vsdfh931cf:30280410
vshan820:30294714
vstes263:30335439
vscorecescf:30445987
pythondataviewer:30285071
vscod805:30301674
binariesv615:30325510
bridge0708:30335490
bridge0723:30353136
cmake_vspar411:30581797
vsaa593:30376534
pythonvs932:30410667
cppdebug:30492333
vsclangdf:30486550
c4g48928:30535728
dsvsc012:30540252
azure-dev_surveyone:30548225
pyindex848cf:30577861
nodejswelcome1cf:30587006
fc301958:30595537
2e4cg342:30602488
gswce1:30612156
3d0df643:30613357
dbltrim-noruby:30604474
f6dab269:30613381

```

</details>

**Comment by jacleggmsr** [comment]
Further information:  uninstalled the Pylint extension, restarted VSCode.  I have `"python.linting.lintOnSave": true` in my settings.json so now when I save a .py file, the python extension runs pylint and it completes without errors (as expected).

**Comment by eleanorjboyd** [comment]
Hello! I wanted to update to say we have received this issue and will be looking into it!

**Comment by jacleggmsr** [comment]
I've just change the title because I think the previous one was bad as it assumed something about the problem that I don't know to be true.  I've had a play around with reproducing this and all I can say is that it seems to be a problem with renaming folders, or moving files around.  I've stripped down my repo to consist of only:

tester.py:
```
from aim import module
from aim2 import module2

module.my_print("Hello, world.")
module2.my_print("Hello, world.")
```
and
aim\module.py:
```
def my_print(x):
    print(x)
```
and
aim2\module.py:
```
def my_print(x):
    print(x)
```
Now, in tester.py I see this error:
`
[{
	"resource": "/c:/Users/jaclegg/src/AIM-Lab/tester.py",
	"owner": "_generated_diagnostic_collection_name_#6",
	"code": "E0401:import-error",
	"severity": 8,
	"message": "Unable to import 'aim'",
	"source": "Pylint",
	"startLineNumber": 2,
	"startColumn": 1,
	"endLineNumber": 2,
	"endColumn": 23
}]
`

But only at the module.my_print("Hello, world.") command, not the module2.my_print("Hello, world.").  The only difference between these two directories is their history:  `aim` has been around for a while on this project and `aim2` is a directory I just made for testing.  So this is probably something to do with some sort of history being saved for the `aim` directory?  Perhaps this is related to the other issue I filed today, #220?

**Comment by jacleggmsr** [comment]
ps. only workaround seems to be to uninstall the extension

**Comment by alisonatwork** [comment]
I am also experiencing this problem. It seemed to be alright a few weeks ago, then it started getting a bit flaky and I needed to close and reopen VS Code every time I wanted it to "refresh", but now it's just straight-up broken no matter how many times I restart. All local imports return E0401 `import-error`.

To "fix" this, I have added an init hook to the project's .pylintrc:

```
init-hook='import sys; sys.path.append(".")'
```

This seems a bit weird, though, because the Python extension itself successfully finds all the imports, it's only Pylint that fails. It seems to me that the Pylint extension should always add the top-level workspace folder as PYTHONPATH (since that appears to be how the Python extension does it) unless explicitly told otherwise.

**Comment by tomvanschaijk** [comment]
I have the same issue. Simply disabled it and switched to flake8. Hopefully this gets resolved soon....

**Comment by karthiknadig** [comment]
In the `vscode-pylint` extension, we run `pylint` with the CWD set to workspace root. So, when trying out the above command, please test it from that location. This will help us narrow down the bit that is causing `pylint` to miss the imports.

**Comment by alisonatwork** [comment]
For me when I run from a raw command line or a Terminal tab inside VS Code, the E0401 appears, because I do not have have PYTHONPATH set. If I manually set PYTHONPATH to the workspace directory and re-run the command, it works as expected.

What confuses me is the inconsistency. The main Python/Pylance extension appears to resolve imports based on the workspace directory as the top-level search path for modules, so it does not show any type hinting errors, or any other errors.  But the Pylint extension does show these errors. And this is a regression from the Pylint support in the old Python extension.

I have done some digging and I found a setting `python.analysis.extraPaths` which was set to `${workspaceFolder}` for me. I think this might have been my (failed) attempt to fix the Pylint extension from a couple weeks ago, because I just removed the setting, restarted VS Code and still Pylint shows E0401 while the Python/Pylance extension still resolves imports perfectly fine.

The only workaround I have found so far is the init-hook, described above.

**Comment by karthiknadig** [comment]
@alisonatwork is your working directory same as workspace folder root?

@DanielNoord Is there a difference in how imports are resolved for `--from-stdin`?

**Comment by alisonatwork** [comment]
Ah, wait, never mind, I just realized I still had PYTHONPATH exported from a previous run. My apologies, it seems `--from-stdin` doesn't make a difference. I'm not sure how our CI runs Pylint directly and manages to resolve imports from the working directory (which is same as the workspace folder).

It seems the core problem for me is not necessarily that the Pylint extension is behaving differently from Pylint itself, but that it's behaving differently from the Python or Pylance extension.

**Comment by DanielNoord** [comment]
> @DanielNoord Is there a difference in how imports are resolved for `--from-stdin`?

Not that I would know of. I would need to look into the specifics, but it seems this has been resolved!
